### PR TITLE
Add CLI command to dump cgroups metadata

### LINKED ISSUE #23359: Expose pod cgroup metadata using CLI
Add a new subcommand (`cilium cgroups list`) to the cilium CLI to expose pod cgroup metadata - https://github.com/cilium/cilium/blob/master/pkg/cgroups/manager/manager.go#L49-L49. 

Here are the steps:
- Extend commands with a new one for cgroups metadata:  https://github.com/cilium/cilium/tree/master/cilium/cmd/ and https://github.com/cilium/cilium/tree/master/cilium/daemon/cmd. 
- Add a new event type (e.g., `podDumpMetadata`) - https://github.com/cilium/cilium/blob/master/pkg/cgroups/manager/manager.go#L28. Add a new method in `pkg/cgroups/manager.go` that sends this event to a channel, and returns the internal metadata result. The method will be invoked by the above cilium CLI handler.

This will be useful to introspect internal state of the cgroups manager (e.g., debugging).

### PR DESCRIPTION
**Pull request by alexkats**
Add CLI command to dump cgroups metadata

Output example for `cilium cgroups ls`:
```
POD NAME                                POD NAMESPACE   CGROUP IDS
cilium-6thwl                            kube-system     3266
konnectivity-agent-78f5595bfb-dk2p8     kube-system     2098
kube-dns-7d75947b86-z2mms               kube-system     2642
                                                        2658
                                                        2626
                                                        2610
kube-dns-autoscaler-747bd9cc7b-nkfx7    kube-system     2018
metrics-server-v0.5.2-fbb9897c5-c7gmz   kube-system     2162
                                                        2146
```

<details>
<summary>Output JSON example for <code>cilium cgroups ls -o json</code></summary>
<p>

```json
[
  {
    "containers": [
      {
        "cgroup-id": 3266,
        "cgroup-path": "/run/cilium/cgroupv2/kubepods/burstable/podc579c6ab-da45-4508-82fb-14b48d3ee506/1a78b0d308e7277155e3c6b75683b9ea3c2f178a0bd9765c422af60c00a75613"
      }
    ],
    "ips": [
      "10.9.88.5"
    ],
    "name": "cilium-6thwl",
    "namespace": "kube-system"
  },
  {
    "containers": [
      {
        "cgroup-id": 2098,
        "cgroup-path": "/run/cilium/cgroupv2/kubepods/burstable/pod007f0d30-9c06-4e12-a7fd-e7465d72acb5/84a87cd21472cb7acf27d265bbe2049213a287eb385cd4c65520213a9397beb7"
      }
    ],
    "ips": [
      "10.16.2.3"
    ],
    "name": "konnectivity-agent-78f5595bfb-dk2p8",
    "namespace": "kube-system"
  },
  {
    "containers": [
      {
        "cgroup-id": 2642,
        "cgroup-path": "/run/cilium/cgroupv2/kubepods/burstable/pod2e6d5e76-b39f-4543-8c15-8dacd3ac222c/0b097ccaad76905810160315cfd2401e0cbe0e70b849e26bd355ef673cab06b3"
      },
      {
        "cgroup-id": 2658,
        "cgroup-path": "/run/cilium/cgroupv2/kubepods/burstable/pod2e6d5e76-b39f-4543-8c15-8dacd3ac222c/0913e89b8b9ec17be214cc37f912e6684e6bbeeb7fb800961e2f6e67e57972c2"
      },
      {
        "cgroup-id": 2626,
        "cgroup-path": "/run/cilium/cgroupv2/kubepods/burstable/pod2e6d5e76-b39f-4543-8c15-8dacd3ac222c/b573fe575973b281c78d8b945087d7eec73161934d0f6e562028d65a143fb17b"
      },
      {
        "cgroup-id": 2610,
        "cgroup-path": "/run/cilium/cgroupv2/kubepods/burstable/pod2e6d5e76-b39f-4543-8c15-8dacd3ac222c/1836b856a7e2aef8f0e0440786a442ac1855a4dcd19fc8608782c7c07dc26879"
      }
    ],
    "ips": [
      "10.16.2.124"
    ],
    "name": "kube-dns-7d75947b86-z2mms",
    "namespace": "kube-system"
  },
  {
    "containers": [
      {
        "cgroup-id": 2018,
        "cgroup-path": "/run/cilium/cgroupv2/kubepods/burstable/pod235e45f7-1c69-4157-a2a1-b89b7d374072/a30919f804bd80612d26ba7e2a0815b72b4a370a92349c23238f0d2dc8b378fa"
      }
    ],
    "ips": [
      "10.16.2.2"
    ],
    "name": "kube-dns-autoscaler-747bd9cc7b-nkfx7",
    "namespace": "kube-system"
  },
  {
    "containers": [
      {
        "cgroup-id": 2162,
        "cgroup-path": "/run/cilium/cgroupv2/kubepods/burstable/pode0112543-4371-4fd7-9c6f-81de14da0aa6/339e31deb3b389e51d2054e84be3a708c84e8e739c156ebc07d24f44708503a1"
      },
      {
        "cgroup-id": 2146,
        "cgroup-path": "/run/cilium/cgroupv2/kubepods/burstable/pode0112543-4371-4fd7-9c6f-81de14da0aa6/0fce30f2f46cd4d344089ef89cd255ab5acf47aade457565a58a1d791ffd5b00"
      }
    ],
    "ips": [
      "10.16.2.4"
    ],
    "name": "metrics-server-v0.5.2-fbb9897c5-c7gmz",
    "namespace": "kube-system"
  }
]
```

</p>
</details>

<details>
<summary>Output YAML example for <code>cilium cgroups ls -o yaml</code></summary>
<p>

```yaml
- containers:
  - cgroupid: 3266
    cgrouppath: /run/cilium/cgroupv2/kubepods/burstable/podc579c6ab-da45-4508-82fb-14b48d3ee506/1a78b0d308e7277155e3c6b75683b9ea3c2f178a0bd9765c422af60c00a75613
  ips:
  - 10.9.88.5
  name: cilium-6thwl
  namespace: kube-system
- containers:
  - cgroupid: 2098
    cgrouppath: /run/cilium/cgroupv2/kubepods/burstable/pod007f0d30-9c06-4e12-a7fd-e7465d72acb5/84a87cd21472cb7acf27d265bbe2049213a287eb385cd4c65520213a9397beb7
  ips:
  - 10.16.2.3
  name: konnectivity-agent-78f5595bfb-dk2p8
  namespace: kube-system
- containers:
  - cgroupid: 2626
    cgrouppath: /run/cilium/cgroupv2/kubepods/burstable/pod2e6d5e76-b39f-4543-8c15-8dacd3ac222c/b573fe575973b281c78d8b945087d7eec73161934d0f6e562028d65a143fb17b
  - cgroupid: 2610
    cgrouppath: /run/cilium/cgroupv2/kubepods/burstable/pod2e6d5e76-b39f-4543-8c15-8dacd3ac222c/1836b856a7e2aef8f0e0440786a442ac1855a4dcd19fc8608782c7c07dc26879
  - cgroupid: 2658
    cgrouppath: /run/cilium/cgroupv2/kubepods/burstable/pod2e6d5e76-b39f-4543-8c15-8dacd3ac222c/0913e89b8b9ec17be214cc37f912e6684e6bbeeb7fb800961e2f6e67e57972c2
  - cgroupid: 2642
    cgrouppath: /run/cilium/cgroupv2/kubepods/burstable/pod2e6d5e76-b39f-4543-8c15-8dacd3ac222c/0b097ccaad76905810160315cfd2401e0cbe0e70b849e26bd355ef673cab06b3
  ips:
  - 10.16.2.124
  name: kube-dns-7d75947b86-z2mms
  namespace: kube-system
- containers:
  - cgroupid: 2018
    cgrouppath: /run/cilium/cgroupv2/kubepods/burstable/pod235e45f7-1c69-4157-a2a1-b89b7d374072/a30919f804bd80612d26ba7e2a0815b72b4a370a92349c23238f0d2dc8b378fa
  ips:
  - 10.16.2.2
  name: kube-dns-autoscaler-747bd9cc7b-nkfx7
  namespace: kube-system
- containers:
  - cgroupid: 2146
    cgrouppath: /run/cilium/cgroupv2/kubepods/burstable/pode0112543-4371-4fd7-9c6f-81de14da0aa6/0fce30f2f46cd4d344089ef89cd255ab5acf47aade457565a58a1d791ffd5b00
  - cgroupid: 2162
    cgrouppath: /run/cilium/cgroupv2/kubepods/burstable/pode0112543-4371-4fd7-9c6f-81de14da0aa6/339e31deb3b389e51d2054e84be3a708c84e8e739c156ebc07d24f44708503a1
  ips:
  - 10.16.2.4
  name: metrics-server-v0.5.2-fbb9897c5-c7gmz
  namespace: kube-system
```

</p>
</details>

Fixes: #23359

Signed-off-by: Alex Katsman <alexkats@google.com>

```release-note
Add CLI command to dump cgroups metadata
```

### CODE REVIEW

**Comment by alexkats** [comment]
Hi, I'm not sure I used a good naming everywhere, so I'd appreciate better ideas :slightly_smiling_face:

**Comment by aditighag** [comment]
Thanks for the PR. I took a brief look at the changes in the manager. 
Could you post a snippet of the new cli subcommand output in the PR/commit description?

**Comment by alexkats** [comment]
Great point, thanks, added examples to the PR description.

File: cilium/cmd/cgroups.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// Copyright Authors of Cilium
+
+package cmd
+
+import "github.com/spf13/cobra"
+
+// cgroupsCmd represents the cgroups command
+var cgroupsCmd = &cobra.Command{
+	Use:   "cgroups",
+	Short: "Manage cgroups",
```
Comment (aditighag) [inline cilium/cmd/cgroups.go]: I would reword the description to "Display cgroup metadata maintained by cilium."

File: cilium/cmd/cgroups_list.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+func printMetadata(w *tabwriter.Writer, metadata *models.CgroupsDumpMetadata) {
+	if metadata == nil {
+		fmt.Fprint(w, "Unable to retrieve cgroups metadata")
+		w.Flush()
+		os.Exit(1)
+	}
+
+	containerMetas := metadata.ContainerMetadatas
+	sort.Slice(containerMetas, func(i, j int) bool { return containerMetas[i].CgroupID < containerMetas[j].CgroupID })
```
Comment (aditighag) [inline cilium/cmd/cgroups_list.go]: Why do we have to sort the data if we are just going to dump it?

File: cilium/cmd/cgroups_list.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		fmt.Fprintf(w, "%s\t%s\t%s\t%s\t\n", cgroupIdTitle, podNameTitle, podNamespaceTitle, podIpsTitle)
+	}
+
+	for _, cm := range metadata.ContainerMetadatas {
+		if pm := cm.PodMetadata; pm != nil {
+			for i, ip := range pm.Ips {
+				if i == 0 {
+					fmt.Fprintf(w, "%d\t%s\t%s\t%s\t\n", cm.CgroupID, pm.Name, pm.Namespace, ip)
+				} else {
+					fmt.Fprintf(w, "\t\t\t%s\t\n", ip)
```
Comment (aditighag) [inline cilium/cmd/cgroups_list.go]: I was wondering if we could anchor the data around pods instead of container cgroup ids. That'll allow us to club ips, cgroup ids. Wdyt?

**Review by aditighag — CHANGES_REQUESTED** [review]
I would also suggest posting the subcommand output snippet to the commit description. Thanks!

File: cilium/cmd/cgroups_list.go (hunk 3 above)
Comment (alexkats) [reply cilium/cmd/cgroups_list.go]: Makes sense to me, thanks, will do it.

File: cilium/cmd/cgroups_list.go (hunk 2 above)
Comment (alexkats) [reply cilium/cmd/cgroups_list.go]: I looked at the endpoint list example, they sorted there. It would be easier to find some specific pods if they are sorted (after I'll do grouping by pod name instead of cgroup id). That said I don't have a very strong preference here, so I can remove the sorting part, if it doesn't make much sense to you.

File: cilium/cmd/cgroups.go (hunk 1 above)
Comment (alexkats) [reply cilium/cmd/cgroups.go]: I'd add this description to the subcommand, as `cilium cgroups` is not necessarily about displaying. And this one probably something like "Manage cgroup metadata maintained by cilium". Wdyt?

Comment (alexkats) [reply cilium/cmd/cgroups.go]: Changed naming for both command and subcommand. Let me know if you think I should name it in a different way.

**Review by qmonnet — APPROVED** [review]
Thanks!

Isn't there a risk of confusion when there are several cgroup IDs and several IPs for a pod? Users believing that the cgroup ID on a given line is somehow associated to the IP on the same line? For example:

```
POD NAME                                POD NAMESPACE   CGROUP IDS   POD IPS
kube-dns-7d75947b86-z2mms               kube-system     2642         10.16.2.124
                                                        2658         10.16.2.125
                                                        2626         10.16.2.126
                                                        2610         10.16.2.127
kube-dns-autoscaler-747bd9cc7b-nkfx7    kube-system     2018         10.16.2.2
```

Not sure how to better display the info, though :thinking:

**Comment by alexkats** [comment]
> Isn't there a risk of confusion when there are several cgroup IDs and several IPs for a pod? Users believing that the cgroup ID on a given line is somehow associated to the IP on the same line?

Yeah, I've been thinking about such a case as well, but didn't find a perfect solution.

Here are a couple of possible ideas:
1. All cgroup ids on the same line (could be a mess if there are a lot of cgroup ids)
```
POD NAME                                POD NAMESPACE   CGROUP IDS               POD IPS
kube-dns-7d75947b86-z2mms               kube-system     2642, 2658, 2626, 2610   10.16.2.124
                                                                                 10.16.2.125
                                                                                 10.16.2.126
                                                                                 10.16.2.127
kube-dns-autoscaler-747bd9cc7b-nkfx7    kube-system     2018                     10.16.2.2
```
2. Maximum 2 cgroup ids on the same line:
```
POD NAME                                POD NAMESPACE   CGROUP IDS    POD IPS
kube-dns-7d75947b86-z2mms               kube-system     2642, 2658,   10.16.2.124
                                                        2626, 2610    10.16.2.125
                                                                      10.16.2.126
                                                                      10.16.2.127
kube-dns-autoscaler-747bd9cc7b-nkfx7    kube-system     2018          10.16.2.2
```
3. Separate cgroup ids by commas, but keep them on separate lines
```
POD NAME                                POD NAMESPACE   CGROUP IDS   POD IPS
kube-dns-7d75947b86-z2mms               kube-system     2642,        10.16.2.124
                                                        2658,        10.16.2.125
                                                        2626,        10.16.2.126
                                                        2610         10.16.2.127
kube-dns-autoscaler-747bd9cc7b-nkfx7    kube-system     2018         10.16.2.2
```

File: cilium/cmd/cgroups_list.go (hunk 2 above)
Comment (aditighag) [reply cilium/cmd/cgroups_list.go]: I don't see any reason to sort the output, so let's drop it for now.

**Review by aditighag — CHANGES_REQUESTED** [review]
The latest output format looks good to me. 

Let's drop pod ips from the output. The main objective of the CLI subcommand is to dump cgroup metadata (and there are other ways to get pod ip information). Ideally, it would've been nice to also output cgroup paths, but they are rather long (especially on systemd env). We can revisit this in future if required. 

Let's also use the new CLI command as part of bugtool collection, so that we have the data as part of sysdumps. Do you mind adding the command to this list - https://github.com/cilium/cilium/blob/master/bugtool/cmd/configuration.go#L361-L361? Please add this in a separate commit. Thanks!

**Comment by alexkats** [comment]
I won't use commas then (preserve the current format) and just drop pod ips. Regarding cgroup paths (and pod ips), they will still be present in the full json or yaml output, so it shouldn't be a problem.

Sure, I'll add it to the bugtool, I'll use json format there, as it contains the complete information including cgroup paths. I see some commands also use json output there.

File: cilium/cmd/cgroups.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+// Copyright Authors of Cilium
+
+package cmd
+
+import "github.com/spf13/cobra"
+
+// cgroupsCmd represents the cgroups command
+var cgroupsCmd = &cobra.Command{
+	Use:   "cgroups",
+	Short: "Manage cgroup metadata maintained by Cilium",
```
Comment (aditighag) [inline cilium/cmd/cgroups.go]: Do you see any use case where users will be "managing" cgroup metadata in some way? If not, we should change the description to "Display cgroup metadata...".

Comment (alexkats) [reply cilium/cmd/cgroups.go]: Well, not right now. I'd just be confused when I see "Display" in the command description, and then suddenly there's a subcommand `list` needed to really display. That's my only concern here, but if you think it's ok this way, I can change the description for the `cgroups` as well.

Comment (aditighag) [reply cilium/cmd/cgroups.go]: That's a fair point. How about changing the description to "Cgroup metadata"? The list command description in `cgroups_list.go` then expands on the description further.

File: cilium/cmd/cgroups_list.go (hunk 2 above)
Comment (alexkats) [reply cilium/cmd/cgroups_list.go]: Ok, removed

File: cilium/cmd/cgroups.go (hunk 4 above)
Comment (alexkats) [reply cilium/cmd/cgroups.go]: Changed it, thanks for the suggestion.

**Review by aditighag — APPROVED** [review]
LGTM... Thanks! Let's give other reviewers a day or two.

**Review by ldelossa — APPROVED** [review]
LGTM. Thanks

**Comment by alexkats** [comment]
Hi, I have 2 questions:
1. Can we kick off a round of tests?
2. AFAIU the only required team to review is the cilium/api, which would be @nathanjsweet in this case. Could you please review this PR?

Thanks!

**Comment by aditighag** [comment]
>Can we kick off a round of tests?

The changes are not covered by e2e tests so was there any specific test you wanted to trigger?

**Comment by alexkats** [comment]
> The changes are not covered by e2e tests so was there any specific test you wanted to trigger?
I'm not sure in this particular case, but in general I thought that for every change you need to run tests in Jenkins (and they are shown as required for the PR). But I'm not an expert here 🙂
